lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -2610,3 +2611,457 @@ SILFunction *swift::lookupPrespecializedSymbol(SILModule &M,
+      auto ProtoDecl = proto->castTo<ProtocolType>()->getDecl();
+      Requirement NewRequirement(RequirementKind::Conformance, GenParam, ProtoDecl->getDeclaredType());
+      Requirements.push_back(NewRequirement);
+    } else if(proto->is<ClassType>()) {
+      auto ClassDecl = proto->castTo<ClassType>()->getDecl();
+      Requirement NewRequirement(RequirementKind::Superclass, GenParam, ClassDecl->getDeclaredType());
+      Requirements.push_back(NewRequirement);
+    } else if (proto->is<ProtocolCompositionType>()) {
+      auto CompType = proto->castTo<ProtocolCompositionType>();
+      recursivelyAddCompositeTypes(CompType, GenParam, Requirements);
+    }
+  }
+}
+
+/// Create the signature for the newly generated protocol constrained generic function.
+CanSILFunctionType ExistentialSpecializerTransform::createExistentialSpecializedFunctionType() {
+
+  CanSILFunctionType FTy = F->getLoweredFunctionType();
+  auto ExpectedFTy = F->getLoweredType().castTo<SILFunctionType>();
+  auto *Mod = F->getModule().getSwiftModule();
+  auto &Ctx = Mod->getASTContext();
+  GenericSignature *NewGenericSig;
+  GenericEnvironment *NewGenericEnv;
+
+  // Form a new generic signature based on the old one.
+  GenericSignatureBuilder Builder(Ctx);
+
+  /// If the original function is generic, then maintain the same.
+  /// Does it have generic structure already ?
+  auto HasGenericSignature = FTy->getGenericSignature() != nullptr;
+  if (HasGenericSignature) {
+    auto OrigGenericSig = FTy->getGenericSignature();
+    /// First, add the old generic signature.
+    Builder.addGenericSignature(OrigGenericSig);
+  }
+
+  /// Original list of parameters
+  SmallVector<SILParameterInfo, 4> params;
+  params.append(ExpectedFTy->getParameters().begin(),
+                ExpectedFTy->getParameters().end());
+
+  int GPIdx = 0;
+  /// Convert the protocol arguments of F to generic ones.
+  for (auto const& IdxIt: ExistentialArgDescriptor) { 
+    int Idx = IdxIt.first;
+    auto &param = params[Idx];
+    auto PType = param.getType();
+
+    assert(PType->is<ProtocolType>() ||
+      PType->is<ProtocolCompositionType>() /*|| PType->is<ClassType>()*/);
+    /// Generate new generic parameter.
+    auto *NewGenericParam = GenericTypeParamType::get(0, GPIdx++, Ctx);
+    Builder.addGenericParameter(NewGenericParam);
+    SmallVector<Requirement, 4> Requirements;
+    if (PType->is<ProtocolType>()) { 
+      auto DeclType = PType->castTo<ProtocolType>()->getDecl();
+      Requirement NewRequirement(RequirementKind::Conformance, NewGenericParam, DeclType->getDeclaredType());
+      Requirements.push_back(NewRequirement);
+    } else { 
+      auto CompType = PType->castTo<ProtocolCompositionType>();
+      recursivelyAddCompositeTypes(CompType, NewGenericParam, Requirements);
+    }
+    auto Source = GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+    for (auto &Req : Requirements) {
+      Builder.addRequirement(Req, Source, Mod);
```

Comment:
Instead of adding requirements to a vector and then adding them to the builder, you can just add them directly to the builder